dll/win32/wdmaud.drv/mmixer.c

```diff
@@ -832,37 +824,36 @@ WdmAudCommitWaveBufferByMMixer(
     Result = GetSoundDeviceType(SoundDevice, &DeviceType);
     SND_ASSERT( Result == MMSYSERR_NOERROR );
 
-    Packet = HeapAlloc(GetProcessHeap(), HEAP_ZERO_MEMORY, sizeof(IO_PACKET));
-    if ( ! Packet )
+    lpHeader = HeapAlloc(GetProcessHeap(), HEAP_ZERO_MEMORY, sizeof(KSSTREAM_HEADER));
+    if ( ! lpHeader )
     {
         /* no memory */
         return MMSYSERR_NOMEM;
     }
 
     /* setup stream packet */
-    Packet->Header.Size = sizeof(KSSTREAM_HEADER);
-    Packet->Header.PresentationTime.Numerator = 1;
-    Packet->Header.PresentationTime.Denominator = 1;
-    Packet->Header.Data = OffsetPtr;
-    Packet->Header.FrameExtent = Length;
-    Packet->hDevice = SoundDeviceInstance->Handle;
-    Packet->Overlap = Overlap;
-    Packet->CompletionRoutine = CompletionRoutine;
-    Packet->IoCtl = (DeviceType == WAVE_OUT_DEVICE_TYPE ? IOCTL_KS_WRITE_STREAM : IOCTL_KS_READ_STREAM);
+    lpHeader->Size = sizeof(KSSTREAM_HEADER);
+    lpHeader->PresentationTime.Numerator = 1;
+    lpHeader->PresentationTime.Denominator = 1;
+    lpHeader->Data = OffsetPtr;
+    lpHeader->FrameExtent = Length;
+    Overlap->CompletionContext = (PVOID)lpHeader;
+    Overlap->OriginalCompletionRoutine = CompletionRoutine;
+    IoCtl = (DeviceType == WAVE_OUT_DEVICE_TYPE ? IOCTL_KS_WRITE_STREAM : IOCTL_KS_READ_STREAM);
 
     if (DeviceType == WAVE_OUT_DEVICE_TYPE)
     {
-        Packet->Header.DataUsed = Length;
+        lpHeader->DataUsed = Length;
     }
 
-    hThread = CreateThread(NULL, 0, IoStreamingThread, (LPVOID)Packet, 0, NULL);
-    if (hThread == NULL)
+    Status = NtDeviceIoControlFile(SoundDeviceInstance->Handle, 
+        NULL, ApcRoutine, NULL, (PIO_STATUS_BLOCK)Overlap, IoCtl,
+        NULL, 0, lpHeader, sizeof(KSSTREAM_HEADER));
+        
+    if (!NT_SUCCESS(Status))
     {
-        /* error */
         return MMSYSERR_ERROR;
     }
-
-    CloseHandle(hThread);
-
+    
```

Comment:
The function to be compared against is WdmAudCommitWaveBufferByLegacy (it also uses WaitForSingleObjectEx)
We cannot wait, that would prevent the queuing of multiple buffers. If we need to wait on completion then we must do it somewhere else.

Reply:
What's concerning is that the APC should only get executed when the thread enters an alertable wait state. So it's not clear when it will ever run.
In your tests, did you see it run? Perhaps you could place a breakpoint there and obtain a backtrace so we can see where it gets called? From that we might be able to judge whether things are already okay or if we need to add a wait somewhere.

Reply:
SoundThreadMain in mmebuddy\thread.c, uses WaitForSingleObjectEx in its main loop.

Reply:
Perfect, thanks for checking that out.